htmresearch/frameworks/nlp/classification_model.py

```diff
@@ -19,335 +19,183 @@
-                                         removeStrings=["[identifier deleted]"],
-                                         correctSpell=True)
-    else:
-      sample = TextPreprocess().tokenize(text)
+    # TODO: Should we specify that distances normalized between 0 and 1?
+    # Currently we use whatever the KNN returns.
 
-    return sample
+    raise NotImplementedError
 
 
-  def tokenize(self, text, preprocess=False):
+  def inferDocument(self, document, returnDetailedResults=False,
+                    sortResults=True):
     """
-    Given a bunch of text (could be several sentences) return a single list
-    containing individual tokens.
+    Run inference on the model with this document and return classification
+    results, sampleIds and distances.  By default this routine will tokenize the
+    document and classify using these tokens. A reset is issued after inference.
+    Repeated sampleIds ARE removed from the results.
 
-    @param text         (str)     A bunch of text.
-    @param preprocess   (bool)    Whether or not to preprocess the text data.
+    @param document (str)     The document to classify
+    @param returnDetailedResults
+                    (bool)    If True will return sampleIds and distances.
+                              This could slow things down depending on the
+                              number of stored patterns.
+    @param sortResults (bool) If true the list of sampleIds and distances
+                              will be sorted in order of increasing distances.
 
-    TODO: get rid of @staticmethod above and then we can remove this method.
+    @return  (numpy array) An array of size numLabels. Position i contains
+                           the likelihood that this token belongs to the i'th
+                           category. An array containing all zeros implies no
+                           decision could be made.
+             (list)        A list of unique sampleIds
+             (numpy array) An array of distances from each stored sample
     """
-    return self.prepText(text, preprocess)
+    # Default implementation, can be overridden
 
+    # For each token run inference on the token and accumulate sum of distances
+    # from this token to all other sampleIds.
+    tokenList = self.tokenize(document)
 
-  def writeOutCategories(self, dirName, comparisons=None, labelRefs=None):
-    """
-    For models which define categories with bitmaps, log the categories (and
-    their relative distances) to a JSON specified by the dirName. The JSON will
-    contain the dict of category bitmaps, and if specified, dicts of label
-    references and category bitmap comparisons.
-    """
-    if not hasattr(self, "categoryBitmaps"):
-      raise TypeError("This model does not have category encodings compatible "
-                      "for logging.")
+    lastTokenIndex = len(tokenList) - 1
+    categoryVotes = numpy.zeros(self.numLabels)
```

Comment:
`numLabels` is used here to init an array with one entry for each possible category label, correct? See my comment in the constructor docstring about this.